# **FRAMEWORK V7 DATASET MAESTRO - MACHINE LEARNING**

## **M0. Carga y Validación del Dataset para Machine Learning**

In [1]:
#==========================================================================================
# SCRIPT 25
# CARGA Y VALIDACIÓN DEL DATASET PARA MACHINE LEARNING
#==========================================================================================
#
# Objetivo
#
# Cargar el Dataset Optimizado generado por el Framework IPML
# y validar su integridad antes del preprocesamiento.
#
# Producto de entrada
#
# dataset_machine_learning.parquet
#
#==========================================================================================

In [2]:
#==========================================================================================
# LIBRERÍAS
#==========================================================================================

import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 200)

In [3]:
#==========================================================================================
# CARGA DEL DATASET
#==========================================================================================

url_dataset = (

"https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C12_MASTER/Dataset_Machine_Learning/dataset_machine_learning.parquet"

)

dataset_ml = pd.read_parquet(url_dataset)

print()

print("="*90)
print("DATASET PARA MACHINE LEARNING")
print("="*90)

print()

print(f"Observaciones : {dataset_ml.shape[0]}")

print(f"Variables     : {dataset_ml.shape[1]}")

display(dataset_ml.head())


DATASET PARA MACHINE LEARNING

Observaciones : 528
Variables     : 9


,Precipitacion_mm,ONI,Radiacion_Solar,Humedad_Relativa,Temp_Min_C,Mes,Velocidad_Viento,irca,VolumenUtilDiarioMasa
0,2.70,0.68,17.78,77.88,8.45,1,1.37,0,0
1,2.22,0.57,17.16,78.83,13.72,2,1.14,0,0
2,5.25,0.57,16.63,80.28,13.52,3,1.09,0,0
3,6.16,0.82,16.56,80.32,12.81,4,1.07,0,0
4,3.27,1.03,16.00,79.57,12.24,5,1.14,0,0


In [4]:
#==========================================================================================
# VALIDACIÓN ESTRUCTURAL
#==========================================================================================

print()

print("="*90)
print("VALIDACIÓN ESTRUCTURAL")
print("="*90)

print()

print("Tipos de datos")

display(dataset_ml.dtypes)


VALIDACIÓN ESTRUCTURAL

Tipos de datos


,0
Precipitacion_mm,float64
ONI,float64
Radiacion_Solar,float64
Humedad_Relativa,float64
Temp_Min_C,float64
Mes,int32
Velocidad_Viento,float64
irca,int64
VolumenUtilDiarioMasa,int64


In [5]:
#==========================================================================================
# VALORES FALTANTES
#==========================================================================================

faltantes = pd.DataFrame({

    "Variable": dataset_ml.columns,

    "Valores_Faltantes": dataset_ml.isna().sum(),

    "Porcentaje":

        (

            dataset_ml.isna().sum()

            /

            len(dataset_ml)

            *

            100

        ).round(2)

})

display(faltantes)

,Variable,Valores_Faltantes,Porcentaje
Precipitacion_mm,Precipitacion_mm,0,0.0
ONI,ONI,0,0.0
Radiacion_Solar,Radiacion_Solar,0,0.0
Humedad_Relativa,Humedad_Relativa,0,0.0
Temp_Min_C,Temp_Min_C,0,0.0
Mes,Mes,0,0.0
Velocidad_Viento,Velocidad_Viento,0,0.0
irca,irca,0,0.0
VolumenUtilDiarioMasa,VolumenUtilDiarioMasa,0,0.0


In [6]:
#==========================================================================================
# VARIABLES CONSTANTES
#==========================================================================================

constantes = dataset_ml.nunique()

constantes = constantes[constantes == 1]

print()

print("="*90)
print("VARIABLES CONSTANTES")
print("="*90)

print()

if len(constantes) == 0:

    print("No existen variables constantes.")

else:

    display(constantes)


VARIABLES CONSTANTES

No existen variables constantes.


In [7]:
#==========================================================================================
# RESUMEN
#==========================================================================================

print()

print("="*90)
print("RESUMEN DEL SCRIPT 25")
print("="*90)

print()

print(f"Observaciones           : {dataset_ml.shape[0]}")

print(f"Variables               : {dataset_ml.shape[1]}")

print(f"Valores faltantes       : {dataset_ml.isna().sum().sum()}")

print(f"Variables constantes    : {(dataset_ml.nunique()==1).sum()}")


RESUMEN DEL SCRIPT 25

Observaciones           : 528
Variables               : 9
Valores faltantes       : 0
Variables constantes    : 0


**¿Cuál será la variable objetivo (Target)?**

| Variable | Tipo |
| :--- | :--- |
| Precipitacion_mm | Predictora |
| ONI | Predictora |
| Radiacion_Solar | Predictora |
| Humedad_Relativa | Predictora |
| Temp_Min_C | Predictora |
| Mes | Predictora |
| Velocidad_Viento | Predictora |
| irca | ¿Predictora o Target? |
| VolumenUtilDiarioMasa | ¿Predictora o Target? |

---

**Escenario 1 (el que considero más fuerte) Predecir el volumen útil diario del embalse**

* Target:

  * VolumenUtilDiarioMasa

* Variables explicativas:

  * Precipitacion_mm
  * ONI
  * Radiacion_Solar
  * Humedad_Relativa
  * Temp_Min_C
  * Mes
  * Velocidad_Viento
  * irca

Este escenario responde preguntas como: ***¿Cómo evolucionará la disponibilidad hídrica?***

Es muy alineado con gestión del recurso hídrico.

---

**Escenario 2 Predecir el IRCA**

* Target:

  * irca

* Variables explicativas:

  * clima
  * volumen
  * estacionalidad

Este escenario responde: **¿Cómo evolucionará la calidad del agua?***

---

**Escenario 3**

El Framework no está diseñado para un único problema.

Está diseñado para seleccionar variables.

Por eso seria crear el Framework independiente del modelo.

Es decir:

* Construir un Dataset Optimizado

  * Modelo Hidrológico (Target: VolumenUtilDiarioMasa)
  * Modelo Calidad del Agua (Targe: irca)

De esta forma el Framework sirve para múltiples problemas.

## **M1. Diagnóstico Estadístico para Machine Learning**


In [8]:
#==========================================================================================
# SCRIPT 26
# DIAGNÓSTICO ESTADÍSTICO PARA MACHINE LEARNING
#==========================================================================================
#
# Objetivo
#
# Evaluar el comportamiento estadístico de las variables
# seleccionadas para determinar la necesidad de aplicar
# transformaciones previas al entrenamiento.
#
#==========================================================================================

In [9]:
print("="*90)
print("RESUMEN ESTADÍSTICO")
print("="*90)

display(dataset_ml.describe().T)

RESUMEN ESTADÍSTICO


,count,mean,std,min,25%,50%,75%,max
Precipitacion_mm,528.0,6.066307,3.708268,0.00,3.0850,5.845,8.3725,18.32
ONI,528.0,0.166364,0.938824,-1.36,-0.5525,-0.015,0.6650,2.77
Radiacion_Solar,528.0,16.586723,1.696319,11.97,15.4575,16.495,17.5025,22.33
Humedad_Relativa,528.0,84.681875,5.181825,65.84,81.5600,86.200,88.5400,92.16
Temp_Min_C,528.0,11.102652,2.474984,4.13,9.1850,11.260,13.1425,16.67
Mes,528.0,6.500000,3.455326,1.00,3.7500,6.500,9.2500,12.00
Velocidad_Viento,528.0,1.092746,0.261388,0.48,0.9000,1.030,1.2600,1.88
irca,528.0,0.090909,0.287752,0.00,0.0000,0.000,0.0000,1.00
VolumenUtilDiarioMasa,528.0,0.083333,0.276648,0.00,0.0000,0.000,0.0000,1.00


In [10]:
print("="*90)
print("ASIMETRÍA")
print("="*90)

asimetria = pd.DataFrame({

    "Variable": dataset_ml.columns,

    "Asimetria": dataset_ml.skew().round(3)

})

display(asimetria)

ASIMETRÍA


,Variable,Asimetria
Precipitacion_mm,Precipitacion_mm,0.657
ONI,ONI,0.891
Radiacion_Solar,Radiacion_Solar,0.436
Humedad_Relativa,Humedad_Relativa,-1.056
Temp_Min_C,Temp_Min_C,-0.216
Mes,Mes,0.000
Velocidad_Viento,Velocidad_Viento,0.680
irca,irca,2.854
VolumenUtilDiarioMasa,VolumenUtilDiarioMasa,3.024


In [11]:
print("="*90)
print("CURTOSIS")
print("="*90)

curtosis = pd.DataFrame({

    "Variable": dataset_ml.columns,

    "Curtosis": dataset_ml.kurt().round(3)

})

display(curtosis)

CURTOSIS


,Variable,Curtosis
Precipitacion_mm,Precipitacion_mm,0.220
ONI,ONI,0.242
Radiacion_Solar,Radiacion_Solar,0.519
Humedad_Relativa,Humedad_Relativa,0.723
Temp_Min_C,Temp_Min_C,-0.850
Mes,Mes,-1.217
Velocidad_Viento,Velocidad_Viento,-0.071
irca,irca,6.170
VolumenUtilDiarioMasa,VolumenUtilDiarioMasa,7.170


In [12]:
#==========================================================================================
# DETECCIÓN DE VALORES ATÍPICOS (IQR)
#==========================================================================================

outliers = []

for columna in dataset_ml.columns:

    # Variables binarias
    if dataset_ml[columna].nunique() <= 2:

        cantidad = np.nan

    else:

        q1 = dataset_ml[columna].quantile(0.25)

        q3 = dataset_ml[columna].quantile(0.75)

        iqr = q3 - q1

        inferior = q1 - 1.5 * iqr

        superior = q3 + 1.5 * iqr

        cantidad = (

            (

                dataset_ml[columna] < inferior

            )

            |

            (

                dataset_ml[columna] > superior

            )

        ).sum()

    outliers.append(

        [

            columna,

            cantidad

        ]

    )

outliers = pd.DataFrame(

    outliers,

    columns=[

        "Variable",

        "Outliers"

    ]

)

print()

print("="*90)
print("DETECCIÓN DE VALORES ATÍPICOS (IQR)")
print("="*90)

display(outliers)


DETECCIÓN DE VALORES ATÍPICOS (IQR)


,Variable,Outliers
0,Precipitacion_mm,7.0
1,ONI,12.0
2,Radiacion_Solar,16.0
3,Humedad_Relativa,6.0
4,Temp_Min_C,0.0
5,Mes,0.0
6,Velocidad_Viento,5.0
7,irca,NaN
8,VolumenUtilDiarioMasa,NaN


In [13]:
#==========================================================================================
# RECOMENDACIÓN PARA PREPROCESAMIENTO
#==========================================================================================

def recomendacion(variable, asimetria):

    # Variables binarias

    if dataset_ml[variable].nunique() <= 2:

        return "Variable binaria. No requiere transformación."

    if abs(asimetria) < 0.5:

        return "No requiere transformación."

    elif abs(asimetria) < 1:

        return "Evaluar estandarización."

    else:

        return "Considerar escalamiento robusto."

In [14]:
diagnostico = asimetria.merge(

    outliers,

    on="Variable"

)

diagnostico["Recomendacion"] = diagnostico.apply(

    lambda fila:

    recomendacion(

        fila["Variable"],

        fila["Asimetria"]

    ),

    axis=1

)

display(diagnostico)

,Variable,Asimetria,Outliers,Recomendacion
0,Precipitacion_mm,0.657,7.0,Evaluar estandarización.
1,ONI,0.891,12.0,Evaluar estandarización.
2,Radiacion_Solar,0.436,16.0,No requiere transformación.
3,Humedad_Relativa,-1.056,6.0,Considerar escalamiento robusto.
4,Temp_Min_C,-0.216,0.0,No requiere transformación.
5,Mes,0.000,0.0,No requiere transformación.
6,Velocidad_Viento,0.680,5.0,Evaluar estandarización.
7,irca,2.854,NaN,Variable binaria. No requiere transformación.
8,VolumenUtilDiarioMasa,3.024,NaN,Variable binaria. No requiere transformación.


**1. Resumen estadístico**

El conjunto de datos está conformado por 528 observaciones y 9 variables previamente seleccionadas mediante el índice IPML.

Se evidencia que:

* No existen valores faltantes.
* Todas las variables presentan suficiente variabilidad.
* Las variables climáticas poseen rangos consistentes con el comportamiento esperado de la cuenca.
* Las variables IRCA y VolumenUtilDiarioMasa presentan naturaleza binaria (0 y 1), por lo que su distribución es diferente al resto de variables continuas.

En consecuencia, el dataset conserva adecuadamente la información necesaria para el entrenamiento de modelos predictivos.

---

**2. Asimetría**

Aquí aparece el primer resultado importante.

| Variable | Interpretación |
| :--- | :--- |
| Radiacion_Solar | prácticamente simétrica |
| Temp_Min_C | prácticamente simétrica |
| Mes | perfectamente simétrica |
| Precipitacion_mm | ligera asimetría positiva |
| ONI | ligera asimetría positiva |
| Velocidad_Viento | ligera asimetría positiva |
| Humedad_Relativa | asimetría negativa |
| irca | fuerte asimetría positiva |
| VolumenUtilDiarioMasa | fuerte asimetría positiva |

Esto tiene completamente sentido

Por ejemplo,

* Precipitación

  * La mayoría de días presentan lluvias bajas y pocos eventos extremos.
  * Eso genera exactamente una cola derecha.
  * Es normal.

* ONI

  * Durante muchos meses permanece cercano a cero.

  * Solo algunos períodos presentan eventos fuertes de Niño o Niña.

  * También es esperado.

* Humedad

  * La humedad normalmente permanece alta.

  * Hay pocos días con humedad baja.

  * Por eso aparece una cola hacia la izquierda.

  * También es correcto.

* IRCA

  * Aquí es donde aparece algo muy interesante.

  * Tiene una asimetría enorme.

  * Pero no porque el dato esté mal.

  * Sino porque: IRCA 0 0 0 0 0 0 1 0 0 0, Es prácticamente una variable binaria. Eso produce automáticamente una gran asimetría.Lo mismo ocurre con VolumenUtilDiarioMasa

---

**3. Curtosis**

La curtosis confirma exactamente lo anterior.

* Las variables irca y VolumenUtilDiarioMasa presentan 6.17 y 7.17 Lo que significa colas muy concentradas y pocos valores diferentes.

* No es un problema. Es una consecuencia de ser binarias.

---

**4. Valores atípicos**

Aquí haría una pequeña mejora metodológica.

* Actualmente el algoritmo detectó IRCA 48 outliers y VolumenUtilDiarioMasa 44 outliers. Pero realmente esos no son outliers.

  * Son valores completamente válidos.

  * El algoritmo IQR supone variables continuas.

  * No binarias.

---
**5. Recomendaciones que se realizaron**

  * Modificar el algoritmo para excluir variables binarias.
  * Actualmente tenemos Considerar transformación logarítmica para IRCA Eso realmente no aplica. Porque una variable binaria 0 1 no debe transformarse logarítmicamente.
  * Cambiar la función abs(asimetria)>1 --> Transformación logarítmica por si la variable es binaria --> No requiere transformación

In [15]:
print("="*90)
print("Exportación")
print("="*90)

diagnostico.to_csv(

    "diagnostico_estadistico_ml.csv",

    index=False,

    encoding="utf-8-sig"

)

diagnostico.to_excel(

    "diagnostico_estadistico_ml.xlsx",

    index=False

)

Exportación


## **M2. Preparación, Transformación y Escalamiento del Dataset para Machine Learning - Experimentos**

Objetivo: Aplicar las transformaciones y técnicas de escalamiento definidas a partir del diagnóstico estadístico, generando un nuevo dataset optimizado para el entrenamiento de modelos de Machine Learning.

**Análisis de la Normalización y Optimización**

El proceso de normalización y optimización se realizó en varias etapas, buscando preparar el dataset de la mejor manera para el entrenamiento de modelos de Machine Learning. A continuación, se detalla el análisis de cada paso:

**1. Diagnóstico Estadístico**

Se realizó un diagnóstico exhaustivo de las variables del dataset para entender su distribución y detectar posibles problemas como asimetría, curtosis y valores atípicos. Los resultados clave fueron:

*   **Resumen Estadístico (`.describe().T`)**: Proporcionó una visión general de la centralidad, dispersión y forma de la distribución de cada variable.
*   **Asimetría (`.skew().round(3)`)**: Se calculó la asimetría para cada variable. Se identificaron variables con asimetría significativa (positiva o negativa), lo cual es crucial para la elección del método de escalado.
    *   `irca` y `VolumenUtilDiarioMasa` mostraron una fuerte asimetría positiva, debido a su naturaleza binaria.
    *   `Precipitacion_mm`, `ONI`, `Velocidad_Viento` mostraron ligera asimetría positiva.
    *   `Humedad_Relativa` mostró asimetría negativa.
*   **Curtosis (`.kurt().round(3)`)**: La curtosis confirmó la forma de las distribuciones, especialmente la concentración de valores en las variables binarias (`irca`, `VolumenUtilDiarioMasa`).
*   **Detección de Valores Atípicos (IQR)**: Se utilizó el método del rango intercuartílico (IQR) para identificar posibles valores atípicos. Se observó que este método identificaba 'outliers' en las variables binarias, lo cual se interpretó como una consecuencia de la naturaleza de estas variables y no como un problema real de datos atípicos, sugiriendo una mejora metodológica para excluir variables binarias de esta detección.

**2. Recomendación para Preprocesamiento**

Basado en el diagnóstico estadístico, se implementó una función (`recomendacion`) para sugerir el tipo de transformación necesario para cada variable:

*   **Variables binarias**: No requieren transformación.
*   **Asimetría `abs(asimetria) < 0.5`**: No requieren transformación.
*   **Asimetría `0.5 <= abs(asimetria) < 1`**: Evaluar estandarización.
*   **Asimetría `abs(asimetria) >= 1`**: Considerar escalamiento robusto (o transformación logarítmica si no es binaria).

Esta función generó un diagnóstico (`diagnostico`) que incluyó una columna de 'Recomendacion' para cada variable, guiando las decisiones de preprocesamiento.

**3. Aplicación Automática del Método de Transformación**

El Framework tomó una decisión automática sobre el método de escalado a aplicar a las variables predictoras (X) basándose en el número de variables con alta asimetría y/o presencia de outliers:

*   **Decisión**: Si el número de variables con asimetría `abs(asimetria) > 1` es igual o superior a 2, O si el número de variables con más de 10 outliers es igual o superior a 2, se opta por `RobustScaler`. En caso contrario, se selecciona `StandardScaler`.
*   **Ejecución**: Para el dataset actual, se determinó `RobustScaler` debido a la presencia de 3 variables muy asimétricas (`Precipitacion_mm`, `Humedad_Relativa`, `VolumenUtilDiarioMasa`) y 2 con outliers (`Precipitacion_mm`, `VolumenUtilDiarioMasa`).
*   **Resultado**: Las variables predictoras (`X`) fueron transformadas utilizando el `RobustScaler`, generando `X_transformado`. Esta transformación es más adecuada para datos que contienen muchos valores atípicos o no siguen una distribución normal, ya que utiliza la mediana y el rango intercuartílico, siendo menos sensible a los extremos.

Finalmente, el dataset transformado (`dataset_transformado`) fue reconstruido, uniendo las variables predictoras escaladas con la variable objetivo sin transformar. Este dataset está ahora optimizado y listo para la siguiente fase de construcción de secuencias temporales para el modelado de Machine Learning.

In [16]:
#==========================================================================================
# CARGA DEL DATASET MAESTRO
#==========================================================================================

import pandas as pd

url_dataset = (

"https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C09_MASTER/Dataset_Maestro_Framework_v04.parquet"

)

dataset = pd.read_parquet(url_dataset)

print()

print("="*90)
print("DATASET MAESTRO")
print("="*90)

print()

print(f"Observaciones : {dataset.shape[0]}")

print(f"Variables     : {dataset.shape[1]}")

display(dataset.head())


DATASET MAESTRO

Observaciones : 528
Variables     : 42


,Nodo,Fecha,Precipitacion_mm,Temp_Max_C,Temp_Min_C,Temp_Media_C,Humedad_Relativa,Velocidad_Viento,Radiacion_Solar,Anio,Mes,ONI,Estado_ENSO,VolumenUtilDiarioMasa,irca,nivel_de_riesgo,POBLACION TOTAL,DENSIDAD POBLACIONAL,INDICE DE DESEMPENO INSTITUCIONAL,ACCESO A AGUA POTABLE ADECUADO,PORCENTAJE DE LA POBLACION CON ACCESO A METODOS DE SANEAMIENTO ADECUADOS,COBERTURA DE ACUEDUCTO URBANO,COBERTURA DE ACUEDUCTO RURAL,COBERTURA DE ALCANTARILLADO URBANO,COBERTURA DE ALCANTARILLADO RURAL,CONTINUIDAD DE ACUEDUCTO URBANO,CALIDAD DEL AGUA,AGUAS RESIDUALES TRATADAS,Nivel_Minimo,CONDUCTIVIDAD ELECTRICA,DEMANDA BIOQUIMICA DE OXIGENO (DBO5),DEMANDA QUIMICA DE OXIGENO (DQO),FOSFORO TOTAL,NITROGENO TOTAL,OXIGENO DISUELTO (OD),SOLIDOS SUSPENDIDOS TOTALES,TEMPERATURA,TURBIDEZ,pH,Version_Dataset,IRCA_Fuente,Volumen_Fuente
0,Bogota_Sabana,2015-01-01,2.70,25.30,8.45,18.26,77.88,1.37,17.78,2015,1,0.68,Niño,0,0,Riesgo bajo,0,1,NaN,1,1,1,1,1,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,V04,Imputado,Imputado
1,Bogota_Sabana,2015-02-01,2.22,26.14,13.72,19.02,78.83,1.14,17.16,2015,2,0.57,Niño,0,0,Riesgo bajo,0,1,NaN,1,1,1,1,1,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,V04,Imputado,Imputado
2,Bogota_Sabana,2015-03-01,5.25,27.56,13.52,19.15,80.28,1.09,16.63,2015,3,0.57,Niño,0,0,Riesgo bajo,0,1,NaN,1,1,1,1,1,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,V04,Imputado,Imputado
3,Bogota_Sabana,2015-04-01,6.16,25.50,12.81,19.07,80.32,1.07,16.56,2015,4,0.82,Niño,0,0,Riesgo bajo,0,1,NaN,1,1,1,1,1,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,V04,Imputado,Imputado
4,Bogota_Sabana,2015-05-01,3.27,25.52,12.24,18.99,79.57,1.14,16.00,2015,5,1.03,Niño,0,0,Riesgo bajo,0,1,NaN,1,1,1,1,1,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,V04,Imputado,Imputado


In [17]:
#==========================================================================================
# CARGA DEL RANKING IPML
#==========================================================================================

url_ipml = (

"https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MASTER/C11_MASTER/Universo_Analitico/ranking_ipml_integracion_dimensiones.csv"

)

ipml = pd.read_csv(url_ipml)

In [ ]:
print(ipml.columns.tolist())

['Ranking', 'Variable', 'Tipo', 'Relaciones_Fuertes', 'Indice_Redundancia', 'Nivel_Redundancia', 'Diagnostico_Redundancia', 'VIF', 'Estado_VIF', 'Diagnostico_VIF', 'Recomendacion_VIF', 'Categoria', 'Importancia', 'Peso', 'Puntaje_Conocimiento_Dominio', '% Cobertura', 'Puntaje_Calidad', 'Puntaje_Redundancia', 'Puntaje_Multicolinealidad', 'IPML', 'Nivel_IPML']


In [18]:
ipml["Estado_Final"] = "No Prioritaria"

ipml.loc[ipml["IPML"] >= 0.80, "Estado_Final"] = "Seleccionada"

ipml.loc[
    (ipml["IPML"] >= 0.70) &
    (ipml["IPML"] < 0.80),
    "Estado_Final"
] = "Pendiente por Cobertura"

## **M3. Variable Objetivo**

In [19]:
# Nivel_Minimo
# VolumenUtilDiarioMasa

DOMINIO = "Gestión Hídrica"

VARIABLE_OBJETIVO = "irca"

VENTANA = 12

HORIZONTE = 1

¿Por qué comenzar con irca?

* No significa que sea la única opción.

* Simplemente sirve como caso de demostración del Framework.

* Mañana cualquier usuario podrá cambiar una única línea:

* VARIABLE_OBJETIVO = "Nivel_Minimo" o VARIABLE_OBJETIVO = "CALIDAD DEL AGUA" y todo el Framework seguirá funcionando.

---

**1. Variables objetivo altamente recomendadas**

Son variables que representan un estado del sistema hídrico y cuya predicción aporta directamente a la toma de decisiones.

| Variable | Justificación |
| :--- | :--- |
| irca | Permite anticipar el riesgo de la calidad del agua para consumo humano. |
| CALIDAD DEL AGUA | Resume el estado general del recurso hídrico y facilita la toma de decisiones. |
| Nivel_Minimo | Fundamental para la gestión de embalses, sequías y abastecimiento. |
| VolumenUtilDiarioMasa | Permite proyectar la disponibilidad de agua almacenada. |
| DEMANDA BIOQUÍMICA DE OXÍGENO (DBO5) | Indicador clave de contaminación orgánica. |
| DEMANDA QUÍMICA DE OXÍGENO (DQO) | Refleja la carga contaminante del agua. |
| OXÍGENO DISUELTO (OD) | Indicador esencial para la salud de ecosistemas acuáticos. |
| pH | Variable crítica para evaluar la calidad del agua. |

**2. Variables objetivo recomendables**

Pueden ser predichas dependiendo del objetivo del estudio.

| Variable | Uso |
| :--- | :--- |
| Turbidez | Calidad del agua. |
| Conductividad Eléctrica | Calidad fisicoquímica. |
| Fósforo Total | Contaminación por nutrientes. |
| Nitrógeno Total | Eutrofización. |
| Sólidos Suspendidos Totales | Calidad del agua superficial. |
| Temperatura del agua | Estado ecológico del recurso. |

**3. Variables que normalmente serían predictoras**

Estas explican el comportamiento del sistema.

* Precipitacion_mm
* Radiacion_Solar
* Humedad_Relativa
* Temp_Max_C
* Temp_Min_C
* Temp_Media_C
* Velocidad_Viento
* ONI
* Mes
* Año

Generalmente no tendría sentido que fueran el objetivo principal del Framework, aunque técnicamente podrían modelarse.

---

**4. Catálogo de Variables Objetivo del Framework
Definición**

El Framework no restringe el proceso de aprendizaje automático a una única variable objetivo. Por el contrario, permite seleccionar diferentes variables de respuesta según el problema de gestión hídrica que se desea abordar.

Las variables objetivo se organizan por dominios temáticos, garantizando que el mismo proceso metodológico pueda reutilizarse en distintos escenarios sin modificar la arquitectura del Framework.

**Dominio 1. Calidad del Agua**

* Objetivo del modelo: Anticipar cambios en la calidad fisicoquímica y microbiológica del agua.

* Variables objetivo posibles:

  | Variable | Prioridad |
  | :--- | :--- |
  | irca | Muy Alta |
  | CALIDAD DEL AGUA | Muy Alta |
  | pH | Muy Alta |
  | OXIGENO DISUELTO (OD) | Muy Alta |
  | DEMANDA BIOQUIMICA DE OXIGENO (DBO5) | Muy Alta |
  | DEMANDA QUIMICA DE OXIGENO (DQO) | Muy Alta |
  | TURBIDEZ | Alta |
  | CONDUCTIVIDAD ELECTRICA | Alta |
  | FOSFORO TOTAL | Alta |
  | NITROGENO TOTAL | Alta |
  | SOLIDOS SUSPENDIDOS TOTALES | Alta |

* Aplicación:

  * Vigilancia ambiental.
  * Alertas tempranas.
  * Calidad del agua para consumo humano.
  * Gestión de cuencas.

**Dominio 2. Gestión del Recurso Hídrico**

* Objetivo: Predecir la disponibilidad del recurso.

* Variables objetivo:

  | Variable | Prioridad |
  | :--- | :--- |
  | Nivel_Minimo | Muy Alta |
  | VolumenUtilDiarioMasa | Muy Alta |

* Aplicación:

  * Gestión de embalses.
  * Operación de acueductos.
  * Planeación frente a sequías.

**Dominio 3. Infraestructura**

* Objetivo: Evaluar el comportamiento de la infraestructura hídrica.

* Variables objetivo:

  | Variable | Prioridad |
  | :--- | :--- |
  | AGUAS RESIDUALES TRATADAS | Muy Alta |
  | CONTINUIDAD DE ACUEDUCTO URBANO | Alta |
  | COBERTURA DE ACUEDUCTO URBANO | Alta |
  | COBERTURA DE ACUEDUCTO RURAL | Alta |
  | COBERTURA DE ALCANTARILLADO URBANO | Alta |
  | COBERTURA DE ALCANTARILLADO RURAL | Alta |

* Aplicación:

  * Planeación de inversiones.
  * Seguimiento institucional.
  * Cobertura de servicios públicos.


**Dominio 4. Gestión Institucional**

* Objetivo: Apoyar la toma de decisiones administrativas.

* Variables objetivo:

* Aplicación:

  * Seguimiento institucional.
  * Evaluación del desempeño.
  * Planeación estratégica.

**Dominio 5. Gestión Social**

* Objetivo: Analizar el acceso de la población al recurso hídrico.

* Variables objetivo:

  | Variable | Prioridad |
  | :--- | :--- |
  | ACCESO A AGUA POTABLE ADECUADO | Muy Alta |
  | PORCENTAJE DE LA POBLACION CON ACCESO A METODOS... | Muy Alta |
  | POBLACION TOTAL | Alta |
  | DENSIDAD POBLACIONAL | Alta |

* Aplicación:

  * Planeación territorial.
  * Políticas públicas.
  * Cobertura de servicios.


In [20]:
#==========================================================================================
# SELECCIÓN AUTOMÁTICA DE VARIABLES PREDICTORAS
#==========================================================================================

# Asegurar que la columna 'Estado_Final' existe en el DataFrame ipml
ipml["Estado_Final"] = "No Prioritaria"
ipml.loc[ipml["IPML"] >= 0.80, "Estado_Final"] = "Seleccionada"
ipml.loc[
    (ipml["IPML"] >= 0.70) &
    (ipml["IPML"] < 0.80),
    "Estado_Final"
] = "Pendiente por Cobertura"

variables_predictoras = (

    ipml

    .loc[

        ipml["Estado_Final"] == "Seleccionada",

        "Variable"

    ]

    .tolist()

)

# La variable objetivo nunca debe estar como predictora

if VARIABLE_OBJETIVO in variables_predictoras:

    variables_predictoras.remove(VARIABLE_OBJETIVO)

print()

print("="*90)
print("VARIABLES PREDICTORAS")
print("="*90)

print()

print(f"Total : {len(variables_predictoras)}")

print()

for v in variables_predictoras:

    print(v)


VARIABLES PREDICTORAS

Total : 15

Precipitacion_mm
ONI
Radiacion_Solar
Humedad_Relativa
Nivel_Minimo
pH
DEMANDA QUIMICA DE OXIGENO (DQO)
Temp_Min_C
AGUAS RESIDUALES TRATADAS
Mes
Velocidad_Viento
VolumenUtilDiarioMasa
CALIDAD DEL AGUA
OXIGENO DISUELTO (OD)
DEMANDA BIOQUIMICA DE OXIGENO (DBO5)


In [21]:
#==========================================================================================
# VALIDACIÓN DE VARIABLES
#==========================================================================================

variables_experimento = variables_predictoras + [VARIABLE_OBJETIVO]

faltantes = [

    v

    for v in variables_experimento

    if v not in dataset.columns

]

if len(faltantes) > 0:

    raise Exception(

        f"Variables inexistentes en el Dataset Maestro: {faltantes}"

    )

print()

print("="*90)
print("VALIDACIÓN")
print("="*90)

print()

print("Todas las variables existen.")


VALIDACIÓN

Todas las variables existen.


In [22]:
#==========================================================================================
# CONSTRUCCIÓN DEL DATASET DEL EXPERIMENTO
#==========================================================================================

dataset_ml = dataset[variables_experimento].copy()

print()

print("="*90)
print("DATASET DEL EXPERIMENTO")
print("="*90)

print()

print(f"Observaciones : {dataset_ml.shape[0]}")

print(f"Variables     : {dataset_ml.shape[1]}")

display(dataset_ml.head())


DATASET DEL EXPERIMENTO

Observaciones : 528
Variables     : 16


,Precipitacion_mm,ONI,Radiacion_Solar,Humedad_Relativa,Nivel_Minimo,pH,DEMANDA QUIMICA DE OXIGENO (DQO),Temp_Min_C,AGUAS RESIDUALES TRATADAS,Mes,Velocidad_Viento,VolumenUtilDiarioMasa,CALIDAD DEL AGUA,OXIGENO DISUELTO (OD),DEMANDA BIOQUIMICA DE OXIGENO (DBO5),irca
0,2.70,0.68,17.78,77.88,NaN,NaN,NaN,8.45,NaN,1,1.37,0,NaN,NaN,NaN,0
1,2.22,0.57,17.16,78.83,NaN,NaN,NaN,13.72,NaN,2,1.14,0,NaN,NaN,NaN,0
2,5.25,0.57,16.63,80.28,NaN,NaN,NaN,13.52,NaN,3,1.09,0,NaN,NaN,NaN,0
3,6.16,0.82,16.56,80.32,NaN,NaN,NaN,12.81,NaN,4,1.07,0,NaN,NaN,NaN,0
4,3.27,1.03,16.00,79.57,NaN,NaN,NaN,12.24,NaN,5,1.14,0,NaN,NaN,NaN,0


## **M4. Validación de Viabilidad del Experimento**

In [23]:
#==========================================================================================
# COBERTURA DE VARIABLES DEL EXPERIMENTO
#==========================================================================================

import pandas as pd

cobertura = pd.DataFrame({

    "Variable": dataset_ml.columns,

    "% Cobertura": (

        dataset_ml.notna().mean() * 100

    ).round(2)

})

print()

print("="*90)
print("VALIDACIÓN DE VIABILIDAD DEL EXPERIMENTO ")
print("="*90)

print()

display(cobertura)


VALIDACIÓN DE VIABILIDAD DEL EXPERIMENTO 



,Variable,% Cobertura
Precipitacion_mm,Precipitacion_mm,100.00
ONI,ONI,100.00
Radiacion_Solar,Radiacion_Solar,100.00
Humedad_Relativa,Humedad_Relativa,100.00
Nivel_Minimo,Nivel_Minimo,16.48
pH,pH,9.47
DEMANDA QUIMICA DE OXIGENO (DQO),DEMANDA QUIMICA DE OXIGENO (DQO),9.28
Temp_Min_C,Temp_Min_C,100.00
AGUAS RESIDUALES TRATADAS,AGUAS RESIDUALES TRATADAS,18.18
Mes,Mes,100.00


In [24]:
#==========================================================================================
# CLASIFICACIÓN DE COBERTURA
#==========================================================================================

def clasificar_cobertura(x):

    if x >= 90:
        return "Excelente"

    elif x >= 70:
        return "Buena"

    elif x >= 50:
        return "Aceptable"

    elif x >= 20:
        return "Baja"

    else:
        return "Insuficiente"

cobertura["Nivel_Cobertura"] = (

    cobertura["% Cobertura"]

    .apply(clasificar_cobertura)

)

display(cobertura)

,Variable,% Cobertura,Nivel_Cobertura
Precipitacion_mm,Precipitacion_mm,100.00,Excelente
ONI,ONI,100.00,Excelente
Radiacion_Solar,Radiacion_Solar,100.00,Excelente
Humedad_Relativa,Humedad_Relativa,100.00,Excelente
Nivel_Minimo,Nivel_Minimo,16.48,Insuficiente
pH,pH,9.47,Insuficiente
DEMANDA QUIMICA DE OXIGENO (DQO),DEMANDA QUIMICA DE OXIGENO (DQO),9.28,Insuficiente
Temp_Min_C,Temp_Min_C,100.00,Excelente
AGUAS RESIDUALES TRATADAS,AGUAS RESIDUALES TRATADAS,18.18,Insuficiente
Mes,Mes,100.00,Excelente


In [25]:
#==========================================================================================
# DECISIÓN DEL FRAMEWORK
#==========================================================================================

UMBRAL_COBERTURA = 50

cobertura["Decision"] = "Excluir"

cobertura.loc[
    cobertura["% Cobertura"] >= UMBRAL_COBERTURA,
    "Decision"
] = "Utilizar"

print()

print("="*90)
print("DECISIÓN DEL FRAMEWORK")
print("="*90)

print()

display(cobertura)


DECISIÓN DEL FRAMEWORK



,Variable,% Cobertura,Nivel_Cobertura,Decision
Precipitacion_mm,Precipitacion_mm,100.00,Excelente,Utilizar
ONI,ONI,100.00,Excelente,Utilizar
Radiacion_Solar,Radiacion_Solar,100.00,Excelente,Utilizar
Humedad_Relativa,Humedad_Relativa,100.00,Excelente,Utilizar
Nivel_Minimo,Nivel_Minimo,16.48,Insuficiente,Excluir
pH,pH,9.47,Insuficiente,Excluir
DEMANDA QUIMICA DE OXIGENO (DQO),DEMANDA QUIMICA DE OXIGENO (DQO),9.28,Insuficiente,Excluir
Temp_Min_C,Temp_Min_C,100.00,Excelente,Utilizar
AGUAS RESIDUALES TRATADAS,AGUAS RESIDUALES TRATADAS,18.18,Insuficiente,Excluir
Mes,Mes,100.00,Excelente,Utilizar


In [26]:
#==========================================================================================
# VALIDACIÓN DE LA VARIABLE OBJETIVO
#==========================================================================================

fila_objetivo = cobertura.loc[
    cobertura["Variable"] == VARIABLE_OBJETIVO
]

cobertura_objetivo = fila_objetivo["% Cobertura"].iloc[0]

print()

print("="*90)
print("VALIDACIÓN DE LA VARIABLE OBJETIVO")
print("="*90)

print()

print(f"Variable objetivo : {VARIABLE_OBJETIVO}")
print(f"Cobertura         : {cobertura_objetivo:.2f}%")

if cobertura_objetivo >= UMBRAL_COBERTURA:

    print()
    print("Resultado : EXPERIMENTO VIABLE")

else:

    raise Exception(

        f"El experimento no es viable. "
        f"La variable objetivo '{VARIABLE_OBJETIVO}' "
        f"solo posee una cobertura del {cobertura_objetivo:.2f}%."

    )


VALIDACIÓN DE LA VARIABLE OBJETIVO

Variable objetivo : irca
Cobertura         : 100.00%

Resultado : EXPERIMENTO VIABLE


## **M5. Construcción del Dataset Final del Experimento**

In [27]:
#==========================================================================================
# CONSTRUCCIÓN DEL DATASET FINAL DEL EXPERIMENTO
#==========================================================================================

# Variables que el Framework decidió utilizar
variables_utilizadas = cobertura.loc[

    cobertura["Decision"] == "Utilizar",

    "Variable"

].tolist()

# Reconstrucción automática del dataset
dataset_ml = dataset_ml[variables_utilizadas].copy()

print()

print("="*90)
print("DATASET FINAL DEL EXPERIMENTO")
print("="*90)

print()

print(f"Observaciones : {dataset_ml.shape[0]}")

print(f"Variables     : {dataset_ml.shape[1]}")

print()

print("Variables utilizadas:")

for variable in dataset_ml.columns:

    print(variable)

print()

display(dataset_ml.head())


DATASET FINAL DEL EXPERIMENTO

Observaciones : 528
Variables     : 9

Variables utilizadas:
Precipitacion_mm
ONI
Radiacion_Solar
Humedad_Relativa
Temp_Min_C
Mes
Velocidad_Viento
VolumenUtilDiarioMasa
irca



,Precipitacion_mm,ONI,Radiacion_Solar,Humedad_Relativa,Temp_Min_C,Mes,Velocidad_Viento,VolumenUtilDiarioMasa,irca
0,2.70,0.68,17.78,77.88,8.45,1,1.37,0,0
1,2.22,0.57,17.16,78.83,13.72,2,1.14,0,0
2,5.25,0.57,16.63,80.28,13.52,3,1.09,0,0
3,6.16,0.82,16.56,80.32,12.81,4,1.07,0,0
4,3.27,1.03,16.00,79.57,12.24,5,1.14,0,0


## **M6. Transformación Automática del Dataset del Experimento**

In [28]:
#==========================================================================================
# SEPARACIÓN DE VARIABLES PREDICTORAS Y OBJETIVO
#==========================================================================================

X = dataset_ml.drop(

    columns=[VARIABLE_OBJETIVO]

)

y = dataset_ml[VARIABLE_OBJETIVO].copy()

print()

print("="*90)
print("VARIABLES PREDICTORAS")
print("="*90)

print()

print(f"Variables predictoras : {X.shape[1]}")

print(f"Variable objetivo     : {VARIABLE_OBJETIVO}")


VARIABLES PREDICTORAS

Variables predictoras : 8
Variable objetivo     : irca


In [29]:
import os
import pandas as pd

# Definir las rutas
url_diagnostico = "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MACHINE_LEARNING/C13_MACHINE_LEARNING/diagnostico_estadistico_ml.csv"
ruta_local_diagnostico = "diagnostico_estadistico_ml.csv"  # Ajusta la ruta si lo guardas en otra carpeta

try:
    # Intentar cargar desde GitHub
    print("Intentando cargar el diagnóstico estadístico desde GitHub...")
    diagnostico = pd.read_csv(url_diagnostico)
    print("¡Carga exitosa desde GitHub!")
except Exception as e:
    print(f"No se pudo cargar desde GitHub ({e}). Intentando cargar desde el archivo local...")
    # Cargar desde la ruta local si falla la web
    diagnostico = pd.read_csv(ruta_local_diagnostico)
    print("¡Carga exitosa desde el archivo local!")

print()
print("=" * 90)
print("DIAGNÓSTICO ESTADÍSTICO")
print("=" * 90)
print()

display(diagnostico)

Intentando cargar el diagnóstico estadístico desde GitHub...
No se pudo cargar desde GitHub (HTTP Error 404: Not Found). Intentando cargar desde el archivo local...
¡Carga exitosa desde el archivo local!

DIAGNÓSTICO ESTADÍSTICO



,Variable,Asimetria,Outliers,Recomendacion
0,Precipitacion_mm,0.657,7.0,Evaluar estandarización.
1,ONI,0.891,12.0,Evaluar estandarización.
2,Radiacion_Solar,0.436,16.0,No requiere transformación.
3,Humedad_Relativa,-1.056,6.0,Considerar escalamiento robusto.
4,Temp_Min_C,-0.216,0.0,No requiere transformación.
5,Mes,0.000,0.0,No requiere transformación.
6,Velocidad_Viento,0.680,5.0,Evaluar estandarización.
7,irca,2.854,NaN,Variable binaria. No requiere transformación.
8,VolumenUtilDiarioMasa,3.024,NaN,Variable binaria. No requiere transformación.


In [30]:
#==========================================================================================
# SELECCIÓN AUTOMÁTICA DEL MÉTODO DE TRANSFORMACIÓN
#==========================================================================================

variables_asimetricas = diagnostico.loc[

    diagnostico["Asimetria"].abs() > 1

].shape[0]

variables_outliers = diagnostico.loc[

    diagnostico["Outliers"] > 10

].shape[0]

if (

    variables_asimetricas >= 2

    or

    variables_outliers >= 2

):

    metodo_transformacion = "RobustScaler"

else:

    metodo_transformacion = "StandardScaler"

print()

print("="*90)
print("DECISIÓN DEL FRAMEWORK")
print("="*90)

print()

print(f"Variables muy asimétricas : {variables_asimetricas}")

print(f"Variables con outliers    : {variables_outliers}")

print()

print(f"Método seleccionado : {metodo_transformacion}")


DECISIÓN DEL FRAMEWORK

Variables muy asimétricas : 3
Variables con outliers    : 2

Método seleccionado : RobustScaler


In [31]:
#==========================================================================================
# APLICACIÓN AUTOMÁTICA DEL MÉTODO DE TRANSFORMACIÓN
#==========================================================================================

from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import RobustScaler

#----------------------------------------
# Separación de variables
#----------------------------------------

X = dataset_ml.drop(

    columns=[VARIABLE_OBJETIVO]

)

y = dataset_ml[VARIABLE_OBJETIVO].copy()

#----------------------------------------
# Selección automática
#----------------------------------------

if metodo_transformacion == "StandardScaler":

    scaler = StandardScaler()

else:

    scaler = RobustScaler()

#----------------------------------------
# Transformación
#----------------------------------------

X_transformado = pd.DataFrame(

    scaler.fit_transform(X),

    columns=X.columns,

    index=X.index

)

print()

print("="*90)
print("TRANSFORMACIÓN AUTOMÁTICA")
print("="*90)

print()

print(f"Método utilizado : {metodo_transformacion}")

print()

display(X_transformado.head())


TRANSFORMACIÓN AUTOMÁTICA

Método utilizado : RobustScaler



,Precipitacion_mm,ONI,Radiacion_Solar,Humedad_Relativa,Temp_Min_C,Mes,Velocidad_Viento,VolumenUtilDiarioMasa
0,-0.594799,0.570842,0.628362,-1.191977,-0.710044,-1.000000,0.944444,0.0
1,-0.685579,0.480493,0.325183,-1.055874,0.621605,-0.818182,0.305556,0.0
2,-0.112530,0.480493,0.066015,-0.848138,0.571068,-0.636364,0.166667,0.0
3,0.059574,0.685832,0.031785,-0.842407,0.391661,-0.454545,0.111111,0.0
4,-0.486998,0.858316,-0.242054,-0.949857,0.247631,-0.272727,0.305556,0.0


In [32]:
#==========================================================================================
# RECONSTRUCCIÓN DEL DATASET TRANSFORMADO
#==========================================================================================

dataset_transformado = X_transformado.copy()

dataset_transformado[VARIABLE_OBJETIVO] = y.values

print()

print("="*90)
print("DATASET TRANSFORMADO")
print("="*90)

print()

print(f"Observaciones : {dataset_transformado.shape[0]}")

print(f"Variables     : {dataset_transformado.shape[1]}")

print()

display(dataset_transformado.head())


DATASET TRANSFORMADO

Observaciones : 528
Variables     : 9



,Precipitacion_mm,ONI,Radiacion_Solar,Humedad_Relativa,Temp_Min_C,Mes,Velocidad_Viento,VolumenUtilDiarioMasa,irca
0,-0.594799,0.570842,0.628362,-1.191977,-0.710044,-1.000000,0.944444,0.0,0
1,-0.685579,0.480493,0.325183,-1.055874,0.621605,-0.818182,0.305556,0.0,0
2,-0.112530,0.480493,0.066015,-0.848138,0.571068,-0.636364,0.166667,0.0,0
3,0.059574,0.685832,0.031785,-0.842407,0.391661,-0.454545,0.111111,0.0,0
4,-0.486998,0.858316,-0.242054,-0.949857,0.247631,-0.272727,0.305556,0.0,0


In [33]:
#==========================================================================================
# EXPORTACIÓN DEL DATASET TRANSFORMADO
#==========================================================================================

dataset_transformado.to_parquet(

    "dataset_machine_learning_transformado.parquet",

    index=False

)

dataset_transformado.to_csv(

    "dataset_machine_learning_transformado.csv",

    index=False,

    encoding="utf-8-sig"

)

print()

print("="*90)
print("EXPORTACIÓN")
print("="*90)

print()

print("Archivos generados:")

print("- dataset_machine_learning_transformado.parquet")

print("- dataset_machine_learning_transformado.csv")


EXPORTACIÓN

Archivos generados:
- dataset_machine_learning_transformado.parquet
- dataset_machine_learning_transformado.csv


In [34]:
import joblib
import os

EXPERIMENTO = "EXP_01"
CARPETA_TRANSFORMACION = f"transformaciones/{EXPERIMENTO}"

os.makedirs(

    CARPETA_TRANSFORMACION,

    exist_ok=True

)

joblib.dump(

    scaler,

    f"{CARPETA_TRANSFORMACION}/scaler.pkl"

)

['transformaciones/EXP_01/scaler.pkl']

## **M7. Construcción de Secuencias Temporales**

In [35]:
import os
import pandas as pd

#==========================================================================================
# CARGA DEL DATASET TRANSFORMADO
#==========================================================================================


# Definir las rutas
url_dataset = (
    "https://raw.githubusercontent.com/jriatiga/FRAMEWORK_V7/refs/heads/main/DATA/MACHINE_LEARNING/C13_MACHINE_LEARNING/Transformaciones/Exp01/dataset_machine_learning_transformado.parquet"
)
ruta_local_dataset = "dataset_machine_learning_transformado.parquet"  # Ajusta la ruta si lo tienes en otra carpeta

try:
    # Intentar cargar desde GitHub
    print("Intentando cargar el dataset transformado desde GitHub...")
    dataset_transformado = pd.read_parquet(url_dataset)
    print("¡Carga exitosa desde GitHub!")
except Exception as e:
    print(f"No se pudo cargar desde GitHub ({e}). Intentando cargar desde el archivo local...")
    # Cargar desde la ruta local si falla la web
    dataset_transformado = pd.read_parquet(ruta_local_dataset)
    print("¡Carga exitosa desde el archivo local!")

print()
print("=" * 90)
print("DATASET TRANSFORMADO")
print("=" * 90)
print()

print(f"Observaciones : {dataset_transformado.shape[0]}")
print(f"Variables     : {dataset_transformado.shape[1]}")

display(dataset_transformado.head())

Intentando cargar el dataset transformado desde GitHub...
No se pudo cargar desde GitHub (HTTP Error 404: Not Found). Intentando cargar desde el archivo local...
¡Carga exitosa desde el archivo local!

DATASET TRANSFORMADO

Observaciones : 528
Variables     : 9


,Precipitacion_mm,ONI,Radiacion_Solar,Humedad_Relativa,Temp_Min_C,Mes,Velocidad_Viento,VolumenUtilDiarioMasa,irca
0,-0.594799,0.570842,0.628362,-1.191977,-0.710044,-1.000000,0.944444,0.0,0
1,-0.685579,0.480493,0.325183,-1.055874,0.621605,-0.818182,0.305556,0.0,0
2,-0.112530,0.480493,0.066015,-0.848138,0.571068,-0.636364,0.166667,0.0,0
3,0.059574,0.685832,0.031785,-0.842407,0.391661,-0.454545,0.111111,0.0,0
4,-0.486998,0.858316,-0.242054,-0.949857,0.247631,-0.272727,0.305556,0.0,0


In [36]:
#==========================================================================================
# CONFIGURACIÓN TEMPORAL
#==========================================================================================

VENTANA = 12

HORIZONTE = 1

print()

print("="*90)
print("CONFIGURACIÓN TEMPORAL")
print("="*90)

print()

print(f"Ventana     : {VENTANA}")

print(f"Horizonte   : {HORIZONTE}")


CONFIGURACIÓN TEMPORAL

Ventana     : 12
Horizonte   : 1


In [37]:
#==========================================================================================
# CONSTRUCCIÓN DE SECUENCIAS TEMPORALES
#==========================================================================================

import numpy as np

# Variables predictoras
X = dataset_transformado.drop(

    columns=[VARIABLE_OBJETIVO]

)

# Variable objetivo
y = dataset_transformado[VARIABLE_OBJETIVO]

# Contenedores
secuencias_X = []

secuencias_y = []

# Construcción de ventanas temporales
for i in range(

    len(dataset_transformado) - VENTANA - HORIZONTE + 1

):

    secuencia = X.iloc[

        i : i + VENTANA

    ].values

    objetivo = y.iloc[

        i + VENTANA + HORIZONTE - 1

    ]

    secuencias_X.append(secuencia)

    secuencias_y.append(objetivo)

print()

print("="*90)
print("SECUENCIAS TEMPORALES")
print("="*90)

print()

print(f"Total de secuencias : {len(secuencias_X)}")


SECUENCIAS TEMPORALES

Total de secuencias : 516


In [38]:
#==========================================================================================
# RESUMEN DE LAS SECUENCIAS
#==========================================================================================

print()

print("="*90)
print("RESUMEN")
print("="*90)

print()

print(

    f"Longitud de cada secuencia : {VENTANA}"

)

print(

    f"Variables predictoras      : {X.shape[1]}"

)

print(

    f"Total de secuencias        : {len(secuencias_X)}"

)


RESUMEN

Longitud de cada secuencia : 12
Variables predictoras      : 8
Total de secuencias        : 516


In [39]:
#==========================================================================================
# PRIMERA SECUENCIA
#==========================================================================================

primera = pd.DataFrame(

    secuencias_X[0],

    columns=X.columns

)

print()

print("="*90)
print("PRIMERA SECUENCIA")
print("="*90)

print()

display(primera)

print()

print("Valor objetivo")

print(

    secuencias_y[0]

)


PRIMERA SECUENCIA



,Precipitacion_mm,ONI,Radiacion_Solar,Humedad_Relativa,Temp_Min_C,Mes,Velocidad_Viento,VolumenUtilDiarioMasa
0,-0.594799,0.570842,0.628362,-1.191977,-0.710044,-1.000000,0.944444,0.0
1,-0.685579,0.480493,0.325183,-1.055874,0.621605,-0.818182,0.305556,0.0
2,-0.112530,0.480493,0.066015,-0.848138,0.571068,-0.636364,0.166667,0.0
3,0.059574,0.685832,0.031785,-0.842407,0.391661,-0.454545,0.111111,0.0
4,-0.486998,0.858316,-0.242054,-0.949857,0.247631,-0.272727,0.305556,0.0
5,-0.341371,1.022587,-0.574572,-1.293696,0.186987,-0.090909,2.361111,0.0
6,-0.343262,1.244353,0.036675,-1.644699,-0.048010,0.090909,2.000000,0.0
7,-0.623168,1.630390,-0.168704,-2.160458,0.138977,0.272727,1.888889,0.0
8,-0.630733,1.852156,0.985330,-2.704871,-0.234997,0.454545,1.666667,0.0
9,-0.254374,1.991786,0.408313,-1.906877,0.419457,0.636364,0.166667,0.0



Valor objetivo
0


In [40]:
#==========================================================================================
# EXPORTACIÓN DE LAS SECUENCIAS TEMPORALES
#==========================================================================================

np.save(

    "secuencias_X.npy",

    np.array(secuencias_X)

)

np.save(

    "secuencias_y.npy",

    np.array(secuencias_y)

)

metadata = pd.DataFrame({

    "Parametro":[

        "Dominio",

        "Variable Objetivo",

        "Ventana",

        "Horizonte",

        "Variables Predictoras",

        "Variables Modelo",

        "Numero de Secuencias"

    ],

    "Valor":[

        DOMINIO,

        VARIABLE_OBJETIVO,

        VENTANA,

        HORIZONTE,

         ";".join(X.columns),

        X.shape[1],

        len(secuencias_X)

    ]

})

metadata.to_csv(

    "metadata_secuencias.csv",

    index=False,

    encoding="utf-8-sig"

)

print()

print("="*90)
print("EXPORTACIÓN")
print("="*90)

print()

print("Archivos generados:")

print("- secuencias_X.npy")

print("- secuencias_y.npy")

print("- metadata_secuencias.csv")


EXPORTACIÓN

Archivos generados:
- secuencias_X.npy
- secuencias_y.npy
- metadata_secuencias.csv


In [41]:
#==========================================================================================
# VALIDACIÓN DEL FRAMEWORK
#==========================================================================================

print()

print("="*90)
print("VALIDACIÓN DEL FRAMEWORK")
print("="*90)

print()

print(f"Dominio                 : {DOMINIO}")

print(f"Variable objetivo       : {VARIABLE_OBJETIVO}")

print(f"Ventana temporal        : {VENTANA}")

print(f"Horizonte               : {HORIZONTE}")

print(f"Variables predictoras   : {X.shape[1]}")

print(f"Secuencias generadas    : {len(secuencias_X)}")

print()

if len(secuencias_X)==len(secuencias_y):

    print("Resultado : VALIDACIÓN EXITOSA")

else:

    raise Exception(

        "Existe inconsistencia entre las secuencias X y y."

    )


VALIDACIÓN DEL FRAMEWORK

Dominio                 : Gestión Hídrica
Variable objetivo       : irca
Ventana temporal        : 12
Horizonte               : 1
Variables predictoras   : 8
Secuencias generadas    : 516

Resultado : VALIDACIÓN EXITOSA


## **M8. Consolidación del Framework de Preparación para Machine Learning**

In [42]:
#==========================================================================================
# CONSOLIDACIÓN DEL FRAMEWORK DE PREPARACIÓN
#==========================================================================================

print()

print("="*90)
print("FRAMEWORK DE PREPARACIÓN PARA MACHINE LEARNING")
print("="*90)

print()

print(f"Dominio                    : {DOMINIO}")
print(f"Variable objetivo          : {VARIABLE_OBJETIVO}")
print(f"Ventana temporal           : {VENTANA}")
print(f"Horizonte                  : {HORIZONTE}")

print()

print(f"Observaciones originales   : {len(dataset_transformado)}")
print(f"Variables predictoras      : {X.shape[1]}")
print(f"Secuencias generadas       : {len(secuencias_X)}")

print()

print("Productos generados:")

print("✔ Dataset transformado")
print("✔ Secuencias temporales")
print("✔ Metadata del experimento")

print()

print("Estado del Framework : FINALIZADO")


FRAMEWORK DE PREPARACIÓN PARA MACHINE LEARNING

Dominio                    : Gestión Hídrica
Variable objetivo          : irca
Ventana temporal           : 12
Horizonte                  : 1

Observaciones originales   : 528
Variables predictoras      : 8
Secuencias generadas       : 516

Productos generados:
✔ Dataset transformado
✔ Secuencias temporales
✔ Metadata del experimento

Estado del Framework : FINALIZADO


In [43]:
#==========================================================================================
# SELECCIÓN DEL MÉTODO DE TRANSFORMACIÓN
#==========================================================================================

def seleccionar_transformacion(

    variables_asimetricas,
    variables_outliers

):

    if variables_asimetricas >= 2:

        return "RobustScaler", RobustScaler()

    else:

        return "StandardScaler", StandardScaler()

In [44]:
METODO_TRANSFORMACION, scaler = seleccionar_transformacion(

    variables_asimetricas,

    variables_outliers

)

In [45]:
#==========================================================================================
# REGISTRO DEL EXPERIMENTO DE PREPARACIÓN
#==========================================================================================

registro = pd.DataFrame({

    "Parametro":[

        "Experimento",
        "Dominio",
        "Variable Objetivo",
        "Ventana Temporal",
        "Horizonte",
        "Variables Predictoras",
        "Observaciones",
        "Número de Secuencias",
        "Método de Transformación"

    ],

    "Valor":[

        "EXP_01",

        DOMINIO,

        VARIABLE_OBJETIVO,

        VENTANA,

        HORIZONTE,

        X.shape[1],

        len(dataset_transformado),

        len(secuencias_X),

        METODO_TRANSFORMACION

    ]

})

print()

print("="*90)
print("REGISTRO DEL EXPERIMENTO")
print("="*90)

print()

display(registro)


REGISTRO DEL EXPERIMENTO



,Parametro,Valor
0,Experimento,EXP_01
1,Dominio,Gestión Hídrica
2,Variable Objetivo,irca
3,Ventana Temporal,12
4,Horizonte,1
5,Variables Predictoras,8
6,Observaciones,528
7,Número de Secuencias,516
8,Método de Transformación,RobustScaler


In [46]:
#==========================================================================================
# EXPORTACIÓN DEL REGISTRO
#==========================================================================================

registro.to_csv(

    "registro_preparacion_exp01.csv",

    index=False,

    encoding="utf-8-sig"

)

registro.to_excel(

    "registro_preparacion_exp01.xlsx",

    index=False

)

print()

print("="*90)
print("EXPORTACIÓN")
print("="*90)

print()

print("Archivos generados")

print()

print("- registro_preparacion_exp01.csv")

print("- registro_preparacion_exp01.xlsx")


EXPORTACIÓN

Archivos generados

- registro_preparacion_exp01.csv
- registro_preparacion_exp01.xlsx
